// react-desktop/src/textInput/osx/textInput.js
'use strict';

var _typeof11 = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _typeof10 = typeof Symbol === "function" && _typeof11(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof11(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof11(obj);
};

var _typeof9 = typeof Symbol === "function" && _typeof10(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof10(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof10(obj);
};

var _typeof8 = typeof Symbol === "function" && _typeof9(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof9(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof9(obj);
};

var _typeof7 = typeof Symbol === "function" && _typeof8(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof8(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof8(obj);
};

var _typeof6 = typeof Symbol === "function" && _typeof7(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof7(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof7(obj);
};

var _typeof5 = typeof Symbol === "function" && _typeof6(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof6(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof6(obj);
};

var _typeof4 = typeof Symbol === "function" && _typeof5(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof5(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof5(obj);
};

var _typeof3 = typeof Symbol === "function" && _typeof4(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof4(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof4(obj);
};

var _typeof2 = typeof Symbol === "function" && _typeof3(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof3(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof3(obj);
};

var _typeof = typeof Symbol === "function" && _typeof2(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof2(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof2(obj);
};

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }return target;
};

var _createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);
    }
  }return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;
  };
}();

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _desktopComponent = require('../../desktopComponent');

var _desktopComponent2 = _interopRequireDefault(_desktopComponent);

var _radium = require('radium');

var _osx10_ = require('./styles/osx10_11');

var _osx10_2 = _interopRequireDefault(_osx10_);

var _label = require('../../label/osx/label');

var _label2 = _interopRequireDefault(_label);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

function _objectWithoutProperties(obj, keys) {
  var target = {};for (var i in obj) {
    if (keys.indexOf(i) >= 0) continue;if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;target[i] = obj[i];
  }return target;
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _possibleConstructorReturn(self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }return call && ((typeof call === "undefined" ? "undefined" : _typeof(call)) === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + (typeof superClass === "undefined" ? "undefined" : _typeof(superClass)));
  }subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } });if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var animation = (0, _radium.keyframes)({
  '0%': {
    opacity: '0',
    borderWidth: '34px',
    top: '-34px',
    left: '-34px'
  },
  '32%': {
    opacity: '0',
    borderRadius: '10px',
    borderWidth: '30px',
    top: '-30px',
    left: '-30px'
  },
  '50%': {
    opacity: '.2',
    borderWidth: '15px',
    top: '-15px',
    left: '-15px'
  },
  '80%': {
    opacity: '.4',
    borderWidth: '9px',
    top: '-9px',
    left: '-9px'
  },
  '90%': {
    opacity: '.9',
    borderWidth: '6px',
    top: '-6px',
    left: '-6px'
  },
  '100%': {
    opacity: '1',
    top: '-3px',
    left: '-3px',
    borderRadius: '4px',
    borderWidth: '3px'
  }
}, 'text-input-focus');

var TextFieldOSX = (_dec = (0, _desktopComponent2.default)(_desktopComponent.PlaceholderStyle, _desktopComponent.Hidden, _desktopComponent.Dimension, _desktopComponent.Margin), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(TextFieldOSX, _Component);

  function TextFieldOSX() {
    _classCallCheck(this, TextFieldOSX);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(TextFieldOSX).apply(this, arguments));
  }

  _createClass(TextFieldOSX, [{
    key: 'getPlaceholderStyle',
    value: function getPlaceholderStyle() {
      return _osx10_2.default.textField[':placeholder'];
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var style = _props.style;
      var width = _props.width;
      var label = _props.label;
      var margin = _props.margin;

      var props = _objectWithoutProperties(_props, ['style', 'width', 'label', 'margin']);

      var componentStyle = _extends({}, style, _osx10_2.default.textField);

      var focusElement = void 0;
      if ((0, _radium.getState)(this.state, 'element', ':focus')) {
        var focusElementStyle = _extends({}, _osx10_2.default.focusElement, {
          animation: 'x .25s linear forwards',
          animationName: animation
        });
        focusElement = _react2.default.createElement('div', { style: focusElementStyle });
      }

      var labelComponent = label ? _react2.default.createElement(_label2.default, { margin: '0 0 3px 0' }, label) : null;

      return _react2.default.createElement('div', _extends({ style: _osx10_2.default.container }, { width: width, margin: margin }), labelComponent, _react2.default.createElement('div', { style: _osx10_2.default.wrapper }, _react2.default.createElement(_radium.StyleRoot, null, focusElement), _react2.default.createElement('input', _extends({
        key: 'element',
        ref: 'element',
        type: 'text',
        style: componentStyle
      }, props))));
    }
  }, {
    key: 'value',
    get: function get() {
      return this.refs.element.value;
    },
    set: function set(value) {
      this.refs.element.value = value;
    }
  }]);

  return TextFieldOSX;
}(_react.Component), _class2.propTypes = {
  label: _react.PropTypes.string
}, _temp)) || _class);
exports.default = TextFieldOSX;
