// react-desktop/src/desktopComponent/windowFocus.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);
    }
  }return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;
  };
}();

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

var WindowFocusComponent = function () {
  function WindowFocusComponent(root) {
    var _this = this;

    _classCallCheck(this, WindowFocusComponent);

    this.windowFocus = function () {
      _this.root.setState({ windowFocused: true });
    };

    this.windowBlur = function () {
      _this.root.setState({ windowFocused: false });
    };

    this.root = root;
    this.root.state.windowFocused = true;
  }

  _createClass(WindowFocusComponent, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (typeof window !== 'undefined') {
        window.addEventListener('focus', this.windowFocus);
        window.addEventListener('blur', this.windowBlur);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (typeof window !== 'undefined') {
        window.removeEventListener('focus', this.windowFocus);
        window.removeEventListener('blur', this.windowBlur);
      }
    }
  }]);

  return WindowFocusComponent;
}();

exports.default = WindowFocusComponent;
