// react-desktop/src/desktopComponent.js
'use strict';

var _typeof12 = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _typeof11 = typeof Symbol === "function" && _typeof12(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof12(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof12(obj);
};

var _typeof10 = typeof Symbol === "function" && _typeof11(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof11(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof11(obj);
};

var _typeof9 = typeof Symbol === "function" && _typeof10(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof10(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof10(obj);
};

var _typeof8 = typeof Symbol === "function" && _typeof9(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof9(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof9(obj);
};

var _typeof7 = typeof Symbol === "function" && _typeof8(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof8(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof8(obj);
};

var _typeof6 = typeof Symbol === "function" && _typeof7(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof7(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof7(obj);
};

var _typeof5 = typeof Symbol === "function" && _typeof6(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof6(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof6(obj);
};

var _typeof4 = typeof Symbol === "function" && _typeof5(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof5(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof5(obj);
};

var _typeof3 = typeof Symbol === "function" && _typeof4(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof4(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof4(obj);
};

var _typeof2 = typeof Symbol === "function" && _typeof3(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof3(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof3(obj);
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Background = exports.Hidden = exports.Alignment = exports.VerticalAlignment = exports.HorizontalAlignment = exports.Padding = exports.Margin = exports.Dimension = exports.PlaceholderStyle = exports.WindowFocus = undefined;

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }return target;
};

var _typeof = typeof Symbol === "function" && _typeof2(Symbol.iterator) === "symbol" ? function (obj) {
  return typeof obj === "undefined" ? "undefined" : _typeof2(obj);
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof2(obj);
};

var _createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);
    }
  }return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;
  };
}();

var _get = function get(object, property, receiver) {
  if (object === null) object = Function.prototype;var desc = Object.getOwnPropertyDescriptor(object, property);if (desc === undefined) {
    var parent = Object.getPrototypeOf(object);if (parent === null) {
      return undefined;
    } else {
      return get(parent, property, receiver);
    }
  } else if ("value" in desc) {
    return desc.value;
  } else {
    var getter = desc.get;if (getter === undefined) {
      return undefined;
    }return getter.call(receiver);
  }
};

exports.default = DesktopComponent;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _color = require('./color');

var _windowFocus = require('./desktopComponent/windowFocus');

var _windowFocus2 = _interopRequireDefault(_windowFocus);

var _placeholderStyle = require('./desktopComponent/placeholderStyle');

var _placeholderStyle2 = _interopRequireDefault(_placeholderStyle);

var _commonStyling = require('./desktopComponent/commonStyling');

var _commonStyling2 = _interopRequireDefault(_commonStyling);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

function _toConsumableArray(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) {
      arr2[i] = arr[i];
    }return arr2;
  } else {
    return Array.from(arr);
  }
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _possibleConstructorReturn(self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }return call && ((typeof call === "undefined" ? "undefined" : _typeof2(call)) === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + (typeof superClass === "undefined" ? "undefined" : _typeof2(superClass)));
  }subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } });if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var WindowFocus = exports.WindowFocus = 'WindowFocus';
var PlaceholderStyle = exports.PlaceholderStyle = 'PlaceholderStyle';
var Dimension = exports.Dimension = function Dimension(defaultWidth, defaultHeight) {
  return ['Dimension', { defaultWidth: defaultWidth, defaultHeight: defaultHeight }];
};
var Margin = exports.Margin = 'Margin';
var Padding = exports.Padding = 'Padding';
var HorizontalAlignment = exports.HorizontalAlignment = 'HorizontalAlignment';
var VerticalAlignment = exports.VerticalAlignment = 'VerticalAlignment';
var Alignment = exports.Alignment = 'Alignment';
var Hidden = exports.Hidden = 'Hidden';
var Background = exports.Background = 'Background';

function ExtendComposedComponent(options, ComposedComponent) {
  var _class, _class2, _temp;

  var Component = (0, _radium2.default)(_class = (_temp = _class2 = function (_ComposedComponent) {
    _inherits(Component, _ComposedComponent);

    function Component(props, context, updater) {
      _classCallCheck(this, Component);

      var visible = props.visible;
      var display = props.display;
      var theme = props.theme;
      var color = props.color;
      var background = props.background;

      var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Component).call(this, props, context, updater));

      _this._components = [];
      _this._params = {};

      _this.context = _this.context || {};
      _this.state = _this.state || {};

      _this.state.visible = _this.state.visible || visible !== false;
      _this.state.display = _this.state.display || display !== false;

      if (!context.theme) {
        _this.context.theme = theme ? theme : 'light';
      }
      _this.state.theme = theme ? theme : _this.context.theme;

      if (!context.color) {
        _this.context.color = color && typeof color !== 'boolean' ? (0, _color.convertColor)(color) : (0, _color.convertColor)('blue');
      }
      _this.state.color = color && typeof color !== 'boolean' ? (0, _color.convertColor)(color) : _this.context.color;

      if (!context.background) {
        _this.context.background = background ? (0, _color.convertColor)(background) : null;
      }
      _this.state.background = background ? (0, _color.convertColor)(background) : _this.context.background;

      _this.initOptions();
      _this.init();
      return _this;
    }

    _createClass(Component, [{
      key: 'initOptions',
      value: function initOptions() {
        var _this2 = this;

        options.forEach(function (item, key) {
          if ((typeof item === 'undefined' ? 'undefined' : _typeof(item)) === 'object' && Object.prototype.toString.call(item) === '[object Array]') {
            options[key] = item[0];
            _this2._params[item[0]] = item[1];
          } else if (typeof item === 'function') {
            var option = item();
            options[key] = option[0];
            _this2._params[option[0]] = option[1];
          }
        });
      }
    }, {
      key: 'init',
      value: function init() {
        if (options.indexOf('WindowFocus') !== -1) {
          this._components = [].concat(_toConsumableArray(this._components), [new _windowFocus2.default(this)]);
        }
        if (options.indexOf('PlaceholderStyle') !== -1) {
          this._components = [].concat(_toConsumableArray(this._components), [new _placeholderStyle2.default(this)]);
        }
        if (options.indexOf('Dimension') !== -1 || options.indexOf('Margin') !== -1 || options.indexOf('Padding') !== -1 || options.indexOf('HorizontalAlignment') !== -1 || options.indexOf('VerticalAlignment') !== -1 || options.indexOf('Alignment') !== -1 || options.indexOf('Hidden') !== -1 || options.indexOf('Background') !== -1) {
          var componentOptions = {
            dimension: options.indexOf('Dimension') !== -1,
            margin: options.indexOf('Margin') !== -1,
            padding: options.indexOf('Padding') !== -1,
            horizontalAlignment: options.indexOf('HorizontalAlignment') !== -1,
            verticalAlignment: options.indexOf('VerticalAlignment') !== -1,
            alignment: options.indexOf('Alignment') !== -1,
            hidden: options.indexOf('Hidden') !== -1,
            background: options.indexOf('Background') !== -1
          };
          Component.propTypes = _extends({}, Component.propTypes, _commonStyling2.default.propTypes(componentOptions));
          this._components = [].concat(_toConsumableArray(this._components), [new _commonStyling2.default(this, componentOptions, this._params, ComposedComponent.styleRefs)]);
        }
      }
    }, {
      key: 'getChildContext',
      value: function getChildContext() {
        var childContext = {
          parent: this,
          color: this.state.color,
          background: typeof this.state.background === 'boolean' ? this.state.color : this.state.background,
          theme: this.state.theme
        };

        if (_get(Object.getPrototypeOf(Component.prototype), 'getChildContext', this)) {
          return _extends({}, childContext, _get(Object.getPrototypeOf(Component.prototype), 'getChildContext', this).call(this));
        }

        return childContext;
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        if (_get(Object.getPrototypeOf(Component.prototype), 'componentDidMount', this)) {
          _get(Object.getPrototypeOf(Component.prototype), 'componentDidMount', this).call(this);
        }

        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = this._components[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var component = _step.value;

            if (component.componentDidMount) {
              component.componentDidMount();
            }
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
    }, {
      key: 'componentDidUpdate',
      value: function componentDidUpdate(prevProps, prevState, prevContext) {
        if (_get(Object.getPrototypeOf(Component.prototype), 'componentDidUpdate', this)) {
          _get(Object.getPrototypeOf(Component.prototype), 'componentDidUpdate', this).call(this, prevProps, prevState, prevContext);
        }

        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = this._components[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var component = _step2.value;

            if (component.componentDidUpdate) {
              component.componentDidUpdate();
            }
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (_get(Object.getPrototypeOf(Component.prototype), 'componentWillUnmount', this)) {
          _get(Object.getPrototypeOf(Component.prototype), 'componentWillUnmount', this).call(this);
        }

        var _iteratorNormalCompletion3 = true;
        var _didIteratorError3 = false;
        var _iteratorError3 = undefined;

        try {
          for (var _iterator3 = this._components[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            var component = _step3.value;

            if (component.componentWillUnmount) {
              component.componentWillUnmount();
            }
          }
        } catch (err) {
          _didIteratorError3 = true;
          _iteratorError3 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion3 && _iterator3.return) {
              _iterator3.return();
            }
          } finally {
            if (_didIteratorError3) {
              throw _iteratorError3;
            }
          }
        }
      }
    }, {
      key: 'shouldComponentUpdate',
      value: function shouldComponentUpdate(nextProps, nextState, nextContext) {
        if (_get(Object.getPrototypeOf(Component.prototype), 'shouldComponentUpdate', this)) {
          return _get(Object.getPrototypeOf(Component.prototype), 'shouldComponentUpdate', this).call(this, nextProps, nextState, nextContext);
        }

        for (var prop in nextContext) {
          if (nextContext.hasOwnProperty(prop)) {
            if (typeof this.context[prop] === 'undefined' || nextContext[prop] !== this.context[prop]) {
              return true;
            }
          }
        }

        for (var _prop in nextProps) {
          if (nextProps.hasOwnProperty(_prop)) {
            if (typeof this.props[_prop] === 'undefined' || nextProps[_prop] !== this.props[_prop]) {
              return true;
            }
          }
        }

        for (var _prop2 in nextState) {
          if (nextState.hasOwnProperty(_prop2)) {
            if (typeof this.state[_prop2] === 'undefined' || nextState[_prop2] !== this.state[_prop2]) {
              return true;
            }
          }
        }

        return false;
      }
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(nextProps, nextContext) {
        if (_get(Object.getPrototypeOf(Component.prototype), 'componentWillReceiveProps', this)) {
          return _get(Object.getPrototypeOf(Component.prototype), 'componentWillReceiveProps', this).call(this, nextProps, nextContext);
        }

        var stateProps = ['visible', 'display', 'theme', 'color', 'background'];
        var stateContext = ['theme', 'color', 'background'];

        var changeState = false;
        var state = {};
        for (var prop in nextProps) {
          if (nextProps.hasOwnProperty(prop) && stateProps.indexOf(prop) !== -1) {
            var value = nextProps[prop];
            if (prop === 'color' && typeof value === 'boolean') {
              continue;
            }
            if (nextProps[prop] !== undefined && value !== this.state[prop]) {
              state[prop] = value;
              changeState = true;
            }
          }
        }

        for (var _prop3 in nextContext) {
          if (nextContext.hasOwnProperty(_prop3) && stateContext.indexOf(_prop3) !== -1) {
            if (nextContext[_prop3] !== undefined && nextContext[_prop3] !== this.state[_prop3]) {
              state[_prop3] = nextContext[_prop3];
              changeState = true;
            }
          }
        }

        if (changeState) {
          this.setState(state);
        }
      }
    }, {
      key: 'render',
      value: function render() {
        for (var _len = arguments.length, params = Array(_len), _key = 0; _key < _len; _key++) {
          params[_key] = arguments[_key];
        }

        var rendered = _get(Object.getPrototypeOf(Component.prototype), 'render', this).call(this, params);

        var _iteratorNormalCompletion4 = true;
        var _didIteratorError4 = false;
        var _iteratorError4 = undefined;

        try {
          for (var _iterator4 = this._components[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
            var component = _step4.value;

            if (component.render) {
              rendered = component.render(rendered);
            }
          }
        } catch (err) {
          _didIteratorError4 = true;
          _iteratorError4 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion4 && _iterator4.return) {
              _iterator4.return();
            }
          } finally {
            if (_didIteratorError4) {
              throw _iteratorError4;
            }
          }
        }

        if (_get(Object.getPrototypeOf(Component.prototype), 'getPlaceholderStyle', this)) {
          rendered = _react2.default.createElement('div', { ref: 'container' }, rendered);
        }

        return rendered;
      }
    }]);

    return Component;
  }(ComposedComponent), _class2.propTypes = _extends({
    children: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.element, _react.PropTypes.array]),
    style: _react.PropTypes.object,
    visible: _react.PropTypes.bool,
    display: _react.PropTypes.bool,
    color: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.bool]),
    theme: _react.PropTypes.string
  }, ComposedComponent.propTypes), _class2.childContextTypes = _extends({
    parent: _react.PropTypes.oneOfType([_react.PropTypes.element, _react.PropTypes.object]),
    color: _react.PropTypes.string,
    background: _react.PropTypes.string,
    theme: _react.PropTypes.string
  }, ComposedComponent.childContextTypes), _class2.contextTypes = _extends({
    parent: _react.PropTypes.oneOfType([_react.PropTypes.element, _react.PropTypes.object]),
    color: _react.PropTypes.string,
    background: _react.PropTypes.string,
    theme: _react.PropTypes.string
  }, ComposedComponent.contextTypes), _temp)) || _class;

  return Component;
}

function DesktopComponent() {
  for (var _len2 = arguments.length, options = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    options[_key2] = arguments[_key2];
  }

  if (options.length === 1 && typeof options[0] === 'function') {
    return ExtendComposedComponent.apply(null, [[], options[0]]);
  }

  return ExtendComposedComponent.bind(null, options);
}
