// electron-compile/lib/packager-cli.js
#!/usr/bin/env node
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.packagerMain = exports.runAsarArchive = exports.packageDirToResourcesDir = undefined;

let packageDirToResourcesDir = exports.packageDirToResourcesDir = (() => {
  var _ref = _asyncToGenerator(function* (packageDir) {
    let appDir = (yield _promise.pfs.readdir(packageDir)).find(function (x) {
      return x.match(/\.app$/i);
    });
    if (appDir) {
      return _path2.default.join(packageDir, appDir, 'Contents', 'Resources', 'app');
    } else {
      return _path2.default.join(packageDir, 'resources', 'app');
    }
  });

  return function packageDirToResourcesDir(_x) {
    return _ref.apply(this, arguments);
  };
})();

let copySmallFile = (() => {
  var _ref2 = _asyncToGenerator(function* (from, to) {
    d(`Copying ${ from } => ${ to }`);

    let buf = yield _promise.pfs.readFile(from);
    yield _promise.pfs.writeFile(to, buf);
  });

  return function copySmallFile(_x2, _x3) {
    return _ref2.apply(this, arguments);
  };
})();

let compileAndShim = (() => {
  var _ref3 = _asyncToGenerator(function* (packageDir) {
    let appDir = yield packageDirToResourcesDir(packageDir);

    d(`Looking in ${ appDir }`);
    for (let entry of yield _promise.pfs.readdir(appDir)) {
      if (entry.match(/^(node_modules|bower_components)$/)) continue;

      let fullPath = _path2.default.join(appDir, entry);
      let stat = yield _promise.pfs.stat(fullPath);

      if (!stat.isDirectory()) continue;

      d(`Executing electron-compile: ${ appDir } => ${ entry }`);
      yield (0, _cli.main)(appDir, [fullPath]);
    }

    d('Copying in es6-shim');
    let packageJson = JSON.parse((yield _promise.pfs.readFile(_path2.default.join(appDir, 'package.json'), 'utf8')));

    let index = packageJson.main || 'index.js';
    packageJson.originalMain = index;
    packageJson.main = 'es6-shim.js';

    yield copySmallFile(_path2.default.join(__dirname, 'es6-shim.js'), _path2.default.join(appDir, 'es6-shim.js'));

    yield _promise.pfs.writeFile(_path2.default.join(appDir, 'package.json'), JSON.stringify(packageJson, null, 2));
  });

  return function compileAndShim(_x4) {
    return _ref3.apply(this, arguments);
  };
})();

let runAsarArchive = exports.runAsarArchive = (() => {
  var _ref4 = _asyncToGenerator(function* (packageDir, asarUnpackDir) {
    let appDir = yield packageDirToResourcesDir(packageDir);

    let asarArgs = ['pack', 'app', 'app.asar'];
    if (asarUnpackDir) {
      asarArgs.push('--unpack-dir', asarUnpackDir);
    }

    var _findExecutableOrGues = findExecutableOrGuess('asar', asarArgs);

    let cmd = _findExecutableOrGues.cmd,
        args = _findExecutableOrGues.args;


    d(`Running ${ cmd } ${ JSON.stringify(args) }`);
    yield (0, _spawnRx.spawnPromise)(cmd, args, { cwd: _path2.default.join(appDir, '..') });
    _rimraf2.default.sync(_path2.default.join(appDir));
  });

  return function runAsarArchive(_x5, _x6) {
    return _ref4.apply(this, arguments);
  };
})();

let packagerMain = exports.packagerMain = (() => {
  var _ref5 = _asyncToGenerator(function* (argv) {
    d(`argv: ${ JSON.stringify(argv) }`);
    argv = argv.splice(2);

    var _splitOutAsarArgument = splitOutAsarArguments(argv);

    let packagerArgs = _splitOutAsarArgument.packagerArgs,
        asarArgs = _splitOutAsarArgument.asarArgs;

    var _findExecutableOrGues2 = findExecutableOrGuess(electronPackager, packagerArgs);

    let cmd = _findExecutableOrGues2.cmd,
        args = _findExecutableOrGues2.args;


    d(`Spawning electron-packager: ${ JSON.stringify(args) }`);
    let packagerOutput = yield (0, _spawnRx.spawnPromise)(cmd, args);
    let packageDirs = parsePackagerOutput(packagerOutput);

    d(`Starting compilation for ${ JSON.stringify(packageDirs) }`);
    for (let packageDir of packageDirs) {
      yield compileAndShim(packageDir);

      if (!asarArgs) continue;

      d('Starting ASAR packaging');
      let asarUnpackDir = null;
      if (asarArgs.length === 2) {
        asarUnpackDir = asarArgs[1];
      }

      yield runAsarArchive(packageDir, asarUnpackDir);
    }
  });

  return function packagerMain(_x7) {
    return _ref5.apply(this, arguments);
  };
})();

exports.splitOutAsarArguments = splitOutAsarArguments;
exports.parsePackagerOutput = parsePackagerOutput;
exports.findExecutableOrGuess = findExecutableOrGuess;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _rimraf = require('rimraf');

var _rimraf2 = _interopRequireDefault(_rimraf);

var _promise = require('./promise');

var _cli = require('./cli');

var _spawnRx = require('spawn-rx');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const d = require('debug-electron')('electron-compile:packager');
const electronPackager = 'electron-packager';

function splitOutAsarArguments(argv) {
  if (argv.find(x => x.match(/^--asar-unpack$/))) {
    throw new Error("electron-compile doesn't support --asar-unpack at the moment, use asar-unpack-dir");
  }

  // Strip --asar altogether
  let ret = argv.filter(x => !x.match(/^--asar/));

  if (ret.length === argv.length) {
    return { packagerArgs: ret, asarArgs: null };
  }

  let indexOfUnpack = ret.findIndex(x => x.match(/^--asar-unpack-dir$/));
  if (indexOfUnpack < 0) {
    return { packagerArgs: ret, asarArgs: [] };
  }

  let unpackArgs = ret.slice(indexOfUnpack, indexOfUnpack + 1);
  let notUnpackArgs = ret.slice(0, indexOfUnpack).concat(ret.slice(indexOfUnpack + 2));

  return { packagerArgs: notUnpackArgs, asarArgs: unpackArgs };
}

function parsePackagerOutput(output) {
  // NB: Yes, this is fragile as fuck. :-/
  console.log(output);
  let lines = output.split('\n');

  let idx = lines.findIndex(x => x.match(/Wrote new app/i));
  if (idx < 1) throw new Error(`Packager output is invalid: ${ output }`);
  lines = lines.splice(idx);

  // Multi-platform case
  if (lines[0].match(/Wrote new apps/)) {
    return lines.splice(1).filter(x => x.length > 1);
  } else {
    return [lines[0].replace(/^.*new app to /, '')];
  }
}

function findExecutableOrGuess(cmdToFind, argsToUse) {
  var _findActualExecutable = (0, _spawnRx.findActualExecutable)(cmdToFind, argsToUse);

  let cmd = _findActualExecutable.cmd,
      args = _findActualExecutable.args;

  if (cmd === electronPackager) {
    d(`Can't find ${ cmdToFind }, falling back to where it should be as a guess!`);
    let cmdSuffix = process.platform === 'win32' ? '.cmd' : '';
    return (0, _spawnRx.findActualExecutable)(_path2.default.resolve(__dirname, '..', '..', '.bin', `${ cmdToFind }${ cmdSuffix }`), argsToUse);
  }

  return { cmd, args };
}

if (process.mainModule === module) {
  packagerMain(process.argv).then(() => process.exit(0)).catch(e => {
    console.error(e.message || e);
    d(e.stack);

    process.exit(-1);
  });
}
